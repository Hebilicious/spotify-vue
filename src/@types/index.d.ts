export type Track = {
    _id: string,
    name: string,
    year: number | null,
    artist: string | null,
    album: string | null,
    duration_s: number,
    imageUri: string,
    type: string,
    uri: string
};

export type Album = {
    type: string,
    _id: string,
    name: string,
    year: number,
    total_tracks: number,
    description: string,
    imageUri: string,
    tracks: Track[]
};

export type PlayList = Album[]