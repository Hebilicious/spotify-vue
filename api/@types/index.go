package api

type Track struct {
	ID       string `json:"_id"`
	Name     string `json:"name"`
	Year     int    `json:"year"`
	Artist   string `json:"artist"`
	Album    string `json:"album"`
	Duration int    `json:"duration_s"`
	ImageUri string `json:"imageUri"`
	Type     string `json:"type"`
	Uri      string `json:"uri"`
}

type Album struct {
	ID          string  `json:"_id"`
	Type        string  `json:"type"`
	Name        string  `json:"name"`
	Year        int     `json:"year"`
	TotalTracks int     `json:"total_tracks"`
	Description string  `json:"description"`
	ImageUri    string  `json:"imageUri"`
	Tracks      []Track `json:"tracks"`
}

type PlayList []Album
