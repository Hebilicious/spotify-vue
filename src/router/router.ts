import DashBoard from "@/views/DashBoard/DashBoard.vue";
import { createRouter, createWebHistory } from "vue-router";

export const routes = [
    { path: '', component: DashBoard },
    { path: '/favourite', component: () => import("@/views/Favourite/Favourite.vue") },
    { path: '/library', component: () => import("@/views/Library/Library.vue") },
    { path: '/album/:_id', component: () => import("@/views/Album/Album.vue") },
]

const router = createRouter({
    history: createWebHistory(),
    routes,

})

export default router;