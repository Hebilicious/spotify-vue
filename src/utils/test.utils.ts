import { VueWrapper } from '@vue/test-utils';
import { expect, it } from 'vitest';

export const testSnapShot = (wrapperGetter: () => VueWrapper) => {
  it('Matches the snapshot', () => {
    expect(wrapperGetter().html()).toMatchSnapshot();
  });
};
