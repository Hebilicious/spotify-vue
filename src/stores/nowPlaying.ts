import { defineStore } from 'pinia';

import { Track } from '@/@types';

export const useNowPlayingStore = defineStore({
  id: 'playback-queue',
  state: () => ({
    playQueue: ([] as Track[]),
    currentIndex: (undefined as unknown as number),
  }),
  getters: {
    getCurrentPlaylist: ({ playQueue }) => playQueue,
    getCurrentPlaying: ({ playQueue, currentIndex }) => currentIndex !== undefined ? playQueue?.[currentIndex] : (undefined as unknown as Track),
  },
  actions: {
    setCurrentPlaylist(newPlayQueue: Track[]) {
      this.$patch({ playQueue: newPlayQueue, currentIndex: 0 });
    },
    next() {
      (this.currentIndex < (this.playQueue.length - 1)) && this.currentIndex++;
    },
    prev() {
      this.currentIndex > 0 && this.currentIndex--;
    },
    clearPlaylist() {
      this.playQueue = [];
      this.currentIndex = (undefined as unknown as number);
    }
  },
});
