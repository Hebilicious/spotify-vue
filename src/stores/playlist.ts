import { PlayList } from '@/@types';
import { getAllPlayLists } from '@/services/playListService';
import { defineStore } from 'pinia';

export const usePlayListStore = defineStore({
    id: 'playlists',
    state: () => ({
        playlists: ([] as PlayList),
    }),
    getters: {
        getPlayLists: ({ playlists }) => playlists,
    },
    actions: {
        async loadPlayLists() {
            this.playlists = await getAllPlayLists();
        }
    },
});

export default usePlayListStore;