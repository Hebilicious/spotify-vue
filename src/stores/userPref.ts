import { Album, PlayList, Track } from '@/@types';
import { defineStore } from 'pinia';
import { useNowPlayingStore } from './nowPlaying'

export const useUserPrefStore = defineStore({
    id: 'user-preferences',
    state: () => ({
        likedSongs: ([] as Track[]),
        likedAlbums: ([] as PlayList),
        volumePref: 0.6
    }),
    getters: {
        getLikedSongs: ({ likedSongs }) => likedSongs,
        isLikedSong: ({ likedSongs }) => {
            const playbackStore = useNowPlayingStore();
            return playbackStore.getCurrentPlaying && likedSongs.find(({ _id }) => _id === playbackStore.getCurrentPlaying._id)
        },
        getLikedAlbums: ({ likedAlbums }) => likedAlbums,
        isLikedAlbum: ({ likedAlbums }) => {
            return ({ _id: albumId }: Album) => likedAlbums.find(({ _id }) => _id === albumId);
        },
        getVolumePref: ({ volumePref }) => volumePref
    },
    actions: {
        toggleLikedSong(track: Track) {
            this.likedSongs = this.isLikedSong
                ? this.likedSongs.filter(({ _id }) => _id !== track._id)
                : [...this.likedSongs, track]
        },
        toggleLikedAlbum(album: Album) {
            this.likedAlbums = this.isLikedAlbum(album)
                ? this.likedAlbums.filter(({ _id }) => _id !== album._id)
                : [...this.likedAlbums, album]
        },
        setVolumePref(volumePref: number) {
            this.volumePref = volumePref;
        }
    },
});

function usePlayListStore() {
    throw new Error('Function not implemented.');
}
