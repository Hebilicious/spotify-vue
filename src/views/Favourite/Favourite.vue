<template>
    <div class="favourite" v-if="getLikedSongs.length > 0">
        <div class="playlist-cover-section">
            <div class="playlist-cover">
                <div class="back-icon">
                    <span class="material-icons" @click="router.back()" data-testid="back-icon">reply_all</span>
                </div>
                <div class="playlist-cover-art">
                    <img :src="getLikedSongs?.[0]?.imageUri" />
                </div>
            </div>
            <div class="playlist-info">
                <h4>PLAYLIST</h4>
                <h1>Liked Songs</h1>
                <h5>{{ getLikedSongs.length }} in List</h5>
            </div>
        </div>
        <div class="playlist">
            <table>
                <thead>
                    <th>#</th>
                    <th>Title</th>
                    <th>Artist</th>
                    <th>
                        <span class="material-icons">schedule</span>
                    </th>
                </thead>
                <div></div>
                <tbody>
                    <tr v-for="(song, index) in getLikedSongs" :key="song._id">
                        <td>{{ index + 1 }}</td>
                        <td>
                            <div class="song-info">
                                <img :src="song.imageUri" width="25px" height="25px" />
                                <div class="artist-info">
                                    <span class="heading">{{ song.name }}</span>
                                    <span class="subtitle">{{ song.artist }}</span>
                                </div>
                            </div>
                        </td>
                        <td>{{ song.artist }}</td>
                        <td>{{ `${(song.duration_s / 60) | 0}:${(song.duration_s | 0) % 60}` }}</td>
                    </tr>
                </tbody>
            </table>
        </div>
    </div>
</template>
  
<script lang="ts" setup>
import { useUserPrefStore } from '@/stores/userPref';
import router from '@/router/router';
import { storeToRefs } from 'pinia';
const { getLikedSongs } = storeToRefs(useUserPrefStore());
</script>
<style lang="scss">
.favourite {
    height: 100%;
    width: 100%;
    .playlist-cover-section {
        height: 30vh;
        max-height: 500px;
        min-height: 340px;
        padding: 1rem;
        display: flex;
        background-color: red;
        background-image: linear-gradient(rgba(0, 0, 0, 0.6) 0, #121212 100%),
            var(--background-noise);
        .playlist-cover {
            display: flex;
            flex-direction: column;
            height: 100%;

            .back-icon {
                flex-grow: 1;
                span {
                    color: white;
                }
            }

            .playlist-cover-art {
                display: flex;
                flex-grow: 2;
                align-items: flex-end;

                img {
                    width: 192px;
                    height: 192px;
                    object-fit: cover;
                    overflow: hidden;
                }
            }
        }

        .playlist-info {
            display: flex;
            flex-direction: column;
            padding: 1.5rem 1.5rem 0 1.5rem;
            justify-content: flex-end;
            color: var(--color-heading);

            h1 {
                font-weight: bold;
            }
        }
    }

    .playlist {
        table {
            width: 100%;
            table-layout: auto;
            border-collapse: collapse;

            thead {
                border-bottom: 1px solid rgba(255, 255, 255, 0.1);

                th {
                    padding: 0.5rem;
                    text-align: left;
                }
            }

            td {
                padding: 0.5rem;
            }

            .song-info {
                display: flex;
                align-items: center;
                grid-column-gap: 1rem;
                img {
                    height: 40px;
                    width: 40px;
                    object-fit: cover;
                }

                .artist-info {
                    display: flex;
                    flex-direction: column;

                    .subtitle {
                        font-size: small;
                    }
                }
            }
        }
    }
}
</style>