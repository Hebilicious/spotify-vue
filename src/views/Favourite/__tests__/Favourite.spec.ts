import { testSnapShot } from "@/utils/test.utils";
import { createTestingPinia } from "@pinia/testing";
import { shallowMount } from "@vue/test-utils";
import { describe, expect, fn, it, vi } from "vitest";
import Favourite from "../Favourite.vue";
import router from '@/router/router';
import { useUserPrefStore } from "@/stores/userPref";
import { Track } from "@/@types";

describe('Favourite', () => {

    vi.mock('@/router/router', () => ({
        default: {
            back: fn()
        }
    }));

    testSnapShot(() => shallowMount(Favourite, {
        global: {
            plugins: [createTestingPinia({
                createSpy: () => fn()
            })]
        }
    }))

    it('calls back when the back icon is clicked', async () => {
        const pinia = createTestingPinia({
            createSpy: () => fn()
        });
        useUserPrefStore().toggleLikedSong((<Partial<Track>>{}) as Track)
        const wrapper = shallowMount(Favourite, {
            global: {
                plugins: [pinia]
            }
        });
        await wrapper.find('[data-testid="back-icon"]').trigger('click');
        expect(router.back).toHaveBeenCalled();
    })
})