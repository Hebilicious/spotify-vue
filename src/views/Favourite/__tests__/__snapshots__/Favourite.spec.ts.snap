// Vitest Snapshot v1

exports[`Favourite > Matches the snapshot 1`] = `"<!--v-if-->"`;
